### PR TITLE
Flutter tool - Don't check for updates when run on a bot

### LINKED ISSUE #76106: Don't run update checks in Flutter tool when CI=true in the environment.
This is somewhat related to https://github.com/flutter/flutter/issues/59533 .

For CI runs there is usually no advantage in being warned that a newer version exists and the update check has real performance implications.

`CI=true` is supported by several (most?) CI systems (including [Travis](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables) and [Github Actions](https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables)).

### PR DESCRIPTION
**Pull request by sigurdm**
Flutter tool. Don't check for updates when CI=true

This fixes https://github.com/flutter/flutter/issues/76106

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Could we use the bot detector instead? That has a few different detection techniques built in

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart:253]: We already have an entire `BotDetector` class that does this:
https://github.com/flutter/flutter/blob/74bd7b6f6db6433783a36f94d9e43173745187f4/packages/flutter_tools/lib/src/base/bot_detector.dart#L53

Comment (sigurdm) [reply packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: I was unaware. Of course there is!
Done

**Comment by sigurdm** [comment]
PTAL. I modified the logic, such that explicitly passing `--version-check` will override the bot and --machine logic.
Not sure if that is the best solution, but it seems most flexible:

* --no-version-check: no version check
* --version-check: version check
* no flag: version check unless --machine or bot.

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:108]: Remove extra line.

File: packages/flutter_tools/test/src/context.dart
Comment (jmagman) [inline packages/flutter_tools/test/src/context.dart:138]:
```suggestion
          BotDetector: () => const AlwaysFalseBotDetector(),
```

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:91]: Add tests for your `explicitVersionCheckPassed ` logic: adding `--version-check` explicitly overrides `--machine` and the bot detection.

**Comment by jmagman** [comment]
You should probably also update the title of this PR from `dont_run_update_check_on_CI` since GitHub merging defaults to using the title as the commit message. 😄

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (sigurdm) [reply packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:108]: Done

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (sigurdm) [reply packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:91]: Good idea! - Done

**Comment by jmagman** [comment]
This LGTM once the merge conflict from ba58ad03 is resolved.  Thanks for the contribution (with tests)!

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:92]: @jonahwilliams Is there an alternative to `MockFlutterVersion`, yet.

File: packages/flutter_tools/test/src/context.dart
Comment (zanderso) [inline packages/flutter_tools/test/src/context.dart:138]: This is probably okay, but I'm wondering why it's being added in this PR?

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:92]: We don't have a good alternative. I think using the existing mock is fine until I write a replacement, since It should be straightforward for me to replace them in one go

File: packages/flutter_tools/test/src/context.dart
Comment (sigurdm) [reply packages/flutter_tools/test/src/context.dart]: It was to change the default behavior.

I instead made the `checks that Flutter installation is up-to-date` test have an explicit `false` bot fake.

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:22]: Presubmit checks are failing because this is reported as an unused import.

**Review by zanderso — APPROVED** [review]
lgtm on green presubmit

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (sigurdm) [reply packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:22]: Removed

**Comment by zanderso** [comment]
It looks like a rebase might be needed to fix the 'flutter_plugins' presubmit check.

**Comment by jonahwilliams** [comment]
Friendly ping @sigurdm , could you pull in the latest master so we can get this landed?

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**
